content/posts/jspi.md

```diff
@@ -0,0 +1,419 @@
+date: 2025-05-13T13:33:51-04:00
+draft: true
+tags: ["announcement"]
+# author: "Me"
+# author: ["Me", "You"] # multiple authors
+showToc: true
+TocOpen: false
+draft: false
+hidemeta: false
+comments: false
+description: "Desc Text."
+# canonicalURL: "https://canonical.url/to/page"
+disableHLJS: true # to disable highlightjs
+disableShare: false
+hideSummary: false
+searchHidden: true
+ShowReadingTime: true
+ShowBreadCrumbs: true
+ShowPostNavLinks: true
+cover:
+    # image: "<image path/url>" # image path/url
+    # alt: "<alt text>" # alt text
+    # caption: "<text>" # display caption under cover
+    relative: false # when using page bundles set this to true
+    hidden: true # only hide on current single page
+---
+
+Writing Python code in the browser, we constantly run into what I call the
+sync/async problem. Namely, most Python code and most libc functions are
+synchronous, but most JavaScript APIs are asynchronous. For example, `urllib`
+offers a synchronous API to make a request, but in JavaScript, the normal way to
+make a request is with the `fetch` API which is asynchronous. Python also has
+APIs like `event_loop.run_until_complete()` which are supposed to block until an
+async task is completed. There was simply no way to do this.
+
+JavaScript Promise integration is a new WebAssembly standard that gives us a way
+to work around this. It allows us to make a call that seems synchronous from
+the perspective of C but is actually asynchronous from the perspective of
+Python.
+
+JavaScript Promise integration became a stage 4 finished proposal on April 8,
+2025. Chrome 137, released May 27th, 2025, [supports JavaScript Promise
+Integration](https://developer.chrome.com/release-notes/137#javascript_promise_integration).
+This will make Pyodide quite a lot more useful.
+
+Thanks to Antonio Cuni for feedback on a draft.
+
+## Pyodide's stack switching API
+
+The function `run_sync` stack switches for the resolution of any Python
+awaitable. For example:
+```py
+from pyodide.ffi import run_sync
+from js import fetch
+
+async def async_http_request(url):
+    resp = await fetch(url)
+    return await resp.text()
+
+def await_async_http_request(url):
+    # Suspend for async_http_request to complete
+    return run_sync(async_http_request(url))
+```
+If we call `await_async_http_request()` from JavaScript normally like
+`await_async_http_request("https://example.com")`, it will raise an error. In
+order to suspend for a promise, we must make a "promising" call into Python, and
+the result will itself be a promise. However, for API compatibility, calling a
+synchronous Python function cannot return a promise. If we want to enable stack
+switching, we need to call it like
+`await_async_http_request.callPromising("https://example.com")`. This returns a
+promise. See the full example
+[here](https://github.com/hoodmane/jspi-blog-examples/blob/main/1-pyodide-api/pyodide_example.mjs).
+
+## Using JSPI directly
```

Comment:
It is nice to explain the details about the JSPI, but most of the Pyodide users will not use JSPI in this way. So how about separating the post into two?

- post 1) Explaining how to utilize JSPI in Pyodide, such as how to enable it in the Chrome or node, real life examples for using`run_sync` or `enableRunUntilComplese`.

- post 2) deep dive into JSPI internals and troubleshooting

Reply:
I wonder if we shouldn't make `enableRunUntilComplete` on by default.

Reply:
I don't know if I have a whole blog post worth of stuff to say about how to use JSPI in Pyodide. It's pretty simple. Maybe I can come up with something...

Reply:
Well I do think it makes more sense as a separate blog post since 0.28 doesn't really change anything in terms of JSPI support. I'll add a bit more explanation about how the API works.